prov/hook/trace/src/hook_trace.c

```diff
@@ -48,6 +53,98 @@
+	if (((ret)>0) && (format != FI_CQ_FORMAT_UNSPEC)) { \
+		FI_TRACE((cq)->domain->fabric->hprov, FI_LOG_CQ, \
+			"addr %lu %s\n", \
+			(uint64_t)(addr), \
+			trace_cq_entry_to_str((format), (buf), ret)); \
+	}
+
+#define MAX_CQSTR_LEN 4096
+char cq_str[MAX_CQSTR_LEN];
+
+static char *
+trace_cq_entry_to_str(const enum fi_cq_format format, void *buf, int num) {
+	int index = 0;
+	int len = MAX_CQSTR_LEN - 1;
+	struct fi_cq_entry *entry = NULL;
+	struct fi_cq_msg_entry *mentry = NULL;
+	struct fi_cq_data_entry *dentry = NULL;
+	struct fi_cq_tagged_entry *tentry = NULL;
+	memset(cq_str, 0, MAX_CQSTR_LEN);
+	switch (format) {
+		case FI_CQ_FORMAT_CONTEXT:
+			entry = (struct fi_cq_entry *)buf;
+			snprintf(cq_str, len, "{ ");
+			for (int i=0; i<num; i++) {
+				index = strlen(cq_str);
+				snprintf(&cq_str[index], len-index, "[ctx %p] ", entry[i].op_context);
+			}
+			index = strlen(cq_str);
+			snprintf(&cq_str[index], len-index, "}");
+			break;
+		case FI_CQ_FORMAT_MSG:
+			mentry = (struct fi_cq_msg_entry *)buf;
+			snprintf(cq_str, len, "{ ");
+			for (int i=0; i<num; i++) {
+				index = strlen(cq_str);
+				if (mentry[i].flags & FI_RECV) {
+					snprintf(&cq_str[index], len-index, "[ctx %p flags 0x%lx len %zu] ",
+						mentry[i].op_context, mentry[i].flags,  mentry[i].len);
+				} else {
+					snprintf(&cq_str[index], len-index, "[ctx %p flags 0x%lx] ",
+						mentry[i].op_context, mentry[i].flags);
+				}
+			}
+			index = strlen(cq_str);
+			snprintf(&cq_str[index], len-index, "}");
+			break;
+		case FI_CQ_FORMAT_DATA:
+			dentry = (struct fi_cq_data_entry *)buf;
+			snprintf(cq_str, len, "{ ");
+			for (int i=0; i<num; i++) {
+				index = strlen(cq_str);
+				if (dentry[i].flags & FI_RECV) {
+					snprintf(&cq_str[index], len-index,
+						"[ctx %p flags 0x%lx len %zu buf %p, data %lu] ",
+						dentry[i].op_context, dentry[i].flags,
+						dentry[i].len, dentry[i].buf,
+						(dentry[i].flags & FI_REMOTE_CQ_DATA)?dentry[i].data:0);
+				} else {
+					snprintf(&cq_str[index], len-index, "[ctx %p flags 0x%lx] ",
+						dentry[i].op_context, dentry[i].flags);
+				}
+			}
+			index = strlen(cq_str);
+			snprintf(&cq_str[index], len-index, "}");
+			break;
+		case FI_CQ_FORMAT_TAGGED:
+			tentry = (struct fi_cq_tagged_entry *)buf;
+			snprintf(cq_str, len, "{ ");
+			for (int i=0; i<num; i++) {
+				index = strlen(cq_str);
+				if (tentry[i].flags & FI_RECV) {
+					snprintf(&cq_str[index], len-index,
+						"[ctx %p flags 0x%lx len %zu buf %p, data %lu tag 0x%lx] ",
+						tentry[i].op_context, tentry[i].flags,
+						tentry[i].len, tentry[i].buf,
+						(tentry[i].flags & FI_REMOTE_CQ_DATA)?tentry[i].data:0,
+						 tentry[i].tag);
+				} else {
+					snprintf(&cq_str[index], len-index, "[ctx %p flags 0x%lx] ",
+						tentry[i].op_context, tentry[i].flags);
+				}
+			}
+			index = strlen(cq_str);
+			snprintf(&cq_str[index], len-index, "}");
+			break;
+		default: // should not go here
+			return NULL;
+	}
+	return cq_str;
+}
```

Comment:
We can reduce the overhead here (avoiding the extra string manipulation calls) by restructuring to:
```
#define TRACE_CQ(mycq, ctx, flags, len, buf, data, tag) \
    FI_TRACE(...blah blah...)

static void trace_cq_context(mycq, buf, count)
{
    struct fi_cq_entry *entry;
    size_t i;

    for (entry = buf, i = 0; i < count; i++) {
        TRACE_CQ(mycq, entry->context, entry->flags, 0, NULL, 0, 0)
        entry = entry + 1;
    }
}

static void trace_cq_msg(mycq, buf, count)
{
    struct fi_cq_msg_entry *entry;
    size_t i;

    for (entry = buf, i = 0; i < count; i++) {
        TRACE_CQ(mycq, entry->context, entry->flags, entry->len, NULL, 0, 0)
        entry = entry + 1;
    }
}

/* repeat for other formats */

static ssize_t trace_cq_read(...)
{
    ret = fi_cq_read(...);
    if (ret > 0)
        mycq->trace(mycq, buf, count);
}
```

We can set mycq->trace during cq creation to reference the correct function.  Although, since we're currently not intercepting cq create from this provider directly, you could also declare trace() as an array of function pointers and access that instead.  E.g.
```
static void (*trace)[FI_CQ_FORMAT_TAGGED] = {trace_cq_context, ...};

static ssize_t trace_cq_read(...)
{
    ...
    trace[mycq->format](mycq, buf, count);
}
```
You could add the flags checks directly to TRACE_CQ macro, or handle those in a common routine that then calls TRACE_CQ.

Reply:
If we call mycq->trace,  ==>  TRACE_CQ ==> FI_TRACE(), the <function: line of code> will point to the trace function for each format, rather than the original cq operations.  This may not be useful for debugging.

Reply:
I'm also trying to limit the string manipulation (snprintf) calls.  We could replace FI_TRACE with direct calls to the logging functions.  FI_TRACE is a wrapper to FI_LOG, which is a macro that calls fi_log_enabled() and fi_log().  We could pass __func__ and __LINE__ into our calls and pass them through.  If fi_log_enabled() returns false, there's nothing we actually need to do anyway.

Reply:
When using fi_log, each call will print the <func:line>.  For an cq operation with more than 1 cq entries,  we may want to concatenate all entries into one printout in the log,  instead of one entry per printout with <func:line>.  Otherwise, it may be hard to tell whether the cq operation return multiple entries or multiple cq operations return a single entry. Will this an issue for debugging? To avoid string manipulation, we can add one fi_log_raw(level, format, ...), if fi_log_enabled(),  it can simply do as "fprintf" on the raw data, not including prefixes.

Reply:
I would print 1 completion per line.  It will be easier to read, and the number of completions that an app can retrieve can be arbitrarily large.  We can have a separate print to display the number of completions found, though that's not a priority.

Don't use fprintf directly, as apps can override the fi_log functionality.